helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -694,16 +701,19 @@ void reset() {
           LOG.info("Reset executor for msgType: " + msgType + ", pool: " + pool);
           shutdownAndAwaitTermination(pool, item);
         }
-
-        if (item.factory() != null) {
-          try {
-            item.factory().reset();
-          } catch (Exception ex) {
-            LOG.error("Failed to reset the factory {} of message type {}.", item.factory().toString(),
-                msgType, ex);
-          }
-        }
       }
+      _hdlrFtyRegistry.values()
+          .parallelStream()
+          .map(MsgHandlerFactoryRegistryItem::factory)
+          .distinct()
+          .filter(Objects::nonNull)
+          .forEach(factory -> {
+            try {
+              factory.reset();
+            } catch (Exception ex) {
+              LOG.error("Failed to reset the factory {}.", factory.toString(), ex);
+            }
+          });
```

Comment:
I would suggest to make parallel reset in a different PR. This requires more thinking. 1) we cannot guarantee there is not dependencies and user are not leveraging dependencies because of sequential reset. 2) if there is a problem for parallel reset, you can just revert reset PR instead of entire PR to revert dedup logic.